unix-domain-socket/src/main/scala/akka/stream/alpakka/unixdomainsocket/scaladsl/UnixDomainSocket.scala

```diff
@@ -0,0 +1,515 @@
+            push(out, grab(in))
+
+          override def onUpstreamFinish(): Unit =
+            ()
+        })
+
+        setHandler(out, new OutHandler {
+          override def onPull(): Unit =
+            if (!isClosed(in) && !hasBeenPulled(in)) pull(in)
+        })
+      }
+  }
+}
+
+/**
+ * Provides Unix Domain Socket functionality to Akka Streams with an interface similar to Akka's Tcp class.
+ */
+class UnixDomainSocket(system: ExtendedActorSystem) extends Extension {
+
+  import UnixDomainSocket._
+
+  implicit val materializer: ActorMaterializer = ActorMaterializer()(system)
+  import system.dispatcher
+
+  private val sel = NativeSelectorProvider.getInstance.openSelector
+
+  private val ioThread = new Thread((() => nioEventLoop(sel)): Runnable, "unix-domain-socket-io")
+  ioThread.start()
+
+  CoordinatedShutdown(system).addTask(CoordinatedShutdown.PhaseServiceStop, "stopUnixDomainSocket") { () =>
+    Future.successful {
+      sel.close() // Not much else that we can do
+      Done
+    }
+  }
+
+  /**
+   * Creates a [[UnixDomainSocket.ServerBinding]] instance which represents a prospective Unix Domain Socket
+   * server binding on the given `endpoint`.
+   *
+   * Please note that the startup of the server is asynchronous, i.e. after materializing the enclosing
+   * [[akka.stream.scaladsl.RunnableGraph]] the server is not immediately available. Only after the materialized future
+   * completes is the server ready to accept client connections.
+   *
+   * @param file      The file to listen on
+   * @param backlog   Controls the size of the connection backlog
+   * @param halfClose
+   *                  Controls whether the connection is kept open even after writing has been completed to the accepted
+   *                  TCP connections.
+   *                  If set to true, the connection will implement the TCP half-close mechanism, allowing the client to
+   *                  write to the connection even after the server has finished writing. The TCP socket is only closed
+   *                  after both the client and server finished writing.
+   *                  If set to false, the connection will immediately closed once the server closes its write side,
+   *                  independently whether the client is still attempting to write. This setting is recommended
+   *                  for servers, and therefore it is the default setting.
+   */
+  def bind(file: File,
+           backlog: Int = 128,
+           halfClose: Boolean = false): Source[IncomingConnection, Future[ServerBinding]] = {
+
+    val (incomingConnectionQueue, incomingConnectionSource) =
+      Source
+        .queue[IncomingConnection](2, OverflowStrategy.backpressure)
+        .prefixAndTail(0)
+        .map {
+          case (_, source) =>
+            source
+              .watchTermination() { (mat, done) =>
+                done
+                  .andThen {
+                    case _ =>
+                      try {
```

Comment:
Nice: No `Try` ;-)

Reply:
I've got a Heiko in my head as I code. :-)